### ISSUE TITLE
[ISSUE]: Drag and Drop

### ISSUE THREAD

**Issue by enkelmedia** [umbraco/Umbraco.CMS.Backoffice#1156]
I've look at drag and drop a lot over the last month for my email package since I need this for the designer. I've tried many different approaches:

* Use "dragulajs"
* Use "sortable.js"
* Build my own
* Build my own again
* Build my own a 3rd time
* Try "sortable.js" again

I've of course looked at the current implementation in Bellissima during my work. In it's current state it's not super reliable and it lacks some of the "smoothness", features and stability of something like "sortable.js".

![drag-and-drop](https://github.com/umbraco/Umbraco.CMS.Backoffice/assets/1782524/837ac889-a6b5-40f2-a9a5-fc32f3a76caa)

In the screenshare we can see that the "ghost" element (which might need some better styling) moves in the list but it's not really following the mouse, its lacking behind. We can also notice that the "auto-scroll"-feature, that seem to exist in the code, does not work.

Here is another example of sorting properties, dragging between containers does not seem to work:

![umb-drag-properties](https://github.com/umbraco/Umbraco.CMS.Backoffice/assets/1782524/17a2d378-7616-445f-9ec0-aabe643a7324)

Moving the "Tabs" or "Groups" does not work either:

![umb-drag-tabs](https://github.com/umbraco/Umbraco.CMS.Backoffice/assets/1782524/c57f98fa-8c74-420a-95a0-b944a61e6271)

As I mentioned in the beginning I've tried hard with implementing my own drag and drop for Newsletter Studio, it's been weeks of work on both trying to get sortablejs, dragula and similar libraries to work and to implement a custom drag and drop.

After initially failing with the libraries I tried my own custom approach for far to long and I decided to go back and try to really understand why sortable.js was causing issues with Lit. It turns out that the problem with the js-libraries are that they are manipulating the DOM while dragging which Lit does not like at all. The DOM needs to be exactly like it was when the drag started for Lit not to be confused. This is because Lit uses [marker comments](https://github.com/lit/lit/blob/273ad4e23b8ec97f1a5015dbf398104f535f9c34/packages/lit-html/src/lit-html.ts#L358) to keep track of elements and to efficiently update the DOM (you probably know this already based on this https://github.com/umbraco/Umbraco.CMS.Backoffice/commit/c58e84d61bf448a0fcbd37e483a5cee99bceaa13).

![image](https://github.com/umbraco/Umbraco.CMS.Backoffice/assets/1782524/28da54ab-4270-4f41-b587-3869429f39c6)

I tried this approach of relying on Lit to update the DOM each time we drag over a element and got this to work quite well but I was missing out on features like animations etc.

I'm not sure why you decided to implement your own drag and drop but I wanted to share my own experience here to avoid re-inventing the wheel on two sides. After many attempts of creating a good custom implementation I reverted back to "sortable.js" and tried again to make it work with Lit - the secret is to make sure to store and restore the DOM to the exact state it was in when the drag started.

I've outlined the approach here: https://github.com/SortableJS/Sortable/issues/2089#issuecomment-1902094024 basically taking a copy of the DOM-structure when the drag starts and then re-applying this when the drag ends.

The result looks something like this:

![ns-drag-v2](https://github.com/umbraco/Umbraco.CMS.Backoffice/assets/1782524/796cccef-d1dc-44de-afc8-1365231b91cb)

Please ignore all the bright colors etc. it's just ways to test different states during the dragging. Notice however that the JSON to the right is the whole model and that each array is controlled by one `drag-container`. So we can sort rows, we can sort/drag controls between columns and clone controls from the toolbox into the designer.

It's two different components:
* drag-container - A reusable component with a slot for the consuming component to insert items
* drag-item - The thing that should be draggable/sortable.

The consuming component will have to bind the array being sorted to the `drag-container` and each item to the `drag-item`.

My implementation does not have all the callbacks and settings that the `sorter.controller.ts` in Bellissima currently have but it supports:

* Containers in different shadowDOMs and drag between them
* `groups` to be able to drag different type of elements on the same page (like properties and property groups when editing a document type) 
* Having `drag handles` inside the `drag-item` to only allow the drag to start at certain elements.
* `cloning` elements (like in the screenshare). 
* Different styles for `ghost` and `drag`-element (with the limitations that HTML5 drag API has on the dragged element.
* Setting to allow dragging between containers OR only allow/not allow sorting within the current container.
* Set the `swapThreshold` (a sortable.js-concept) for different containers.

Basically any feature of sortable.js could be used: https://sortablejs.github.io/Sortable/

I basically posted this issue to share my experience with this and to put focus on the question. Since I'm not aware of the reasons why a custom implementation was made I wanted to post this to make you aware that using sortable.js would probably be a viable option. 

I'm of course willing to share the code and collaborate on this if there is interest. If not - no hard feelings =D

Let me know =D

**Comment by nielslyngsoe** [comment]
Hi @enkelmedia 

We have our own Sorter, it was changed a bit lately, I will recommend looking at some of the Pickers or the Block List Editor.

But we should have a look at this at one point. I'll assign my self for now.

**Comment by enkelmedia** [comment]
Hi!

I'll have a look at Pickers and Block List Editor as well =D

Just to be clear: My intention was not to say that something needs to change or that the current "Sorter" would not be a viable option - I just posted this to share my experience around it and offer an alternative if you start to feel pain with the custom one.
